### DISCUSSION TITLE
Proper way to test App with dynamically configure sqlalchemy backend

### DISCUSSION THREAD

**Discussion by nerdstrike** [fastapi/fastapi#6508] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Useful stuff condensed into app.py
app = FastAPI()

engine = create_async_engine(
    config['DB_URL'], future=True
)
Base.metadata.schema = config['DB_SCHEMA']
session_factory = sessionmaker(
    engine, expire_on_commit=False, class_=AsyncSession
)

async def get_DbAccessor():
    'Provides a hook for fastapi to Depend on a DB session in each route'
    async with session_factory() as session:
        async with session.begin():
            yield DataAccessClass(session)

@app.get('/', ...)
async def get_stuff_from_db(db=Depends(get_DbAccessor)):
    things = await db.execute(select(Thing))
    return things.scalars().all()

-----
# fixture.py
@pytest_asyncio.fixture
def insert_some_test_data(session):
    session.add_all([Thing(property='important')])

@pytest_asyncio.fixture
async def async_session():
    engine = create_async_engine(
        'sqlite+aiosqlite:///:memory:', future=True
    )
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
        await conn.run_sync(insert_some_test_data)

    SessionFactory = sqlalchemy.orm.sessionmaker(
        engine, expire_on_commit=False, class_=AsyncSession
    )
    session = SessionFactory()
    yield session
    await session.close()
    await engine.dispose()

@pytest_asyncio.fixture()
async def fastapi_testclient(async_session) -> TestClient:
    async with async_session.begin():
        def _get_db_override():
            return DataAccessClass(async_session)

        app.dependency_overrides[get_DbAccessor] = _get_db_override
        yield TestClient(app)
----
# test.py

def test_test(fastapi_testclient):
    # Here the app connects to $ENV{DB_URL}, but then the dependency calls access sqlite
    assert fastapi_testclient.get('/')
```

### Description

The above skeletal test+app is giving me trouble and I've been all the way down the rabbit hole. I want to be able to test a different DB backend automatically/configure it for several backends in CI.

Problem 1: The tests require that `$ENV` contains `DB_URL` even though it isn't needed. There is modest risk of connecting to a production DB by accident.

Problem 2: It is irritating and fragile to remember to `export DB_URL='test-db'` before invoking pytest. I can use something like `pytest-env` instead, but this will be invisible to developers and fiddly for testing with different backends in CI testing.

I'm struggling to see a way past the extensive use of singletons in the FastAPI docs. The database configuration and connection happens at import time, and I cannot find a nice way to inject a test `DB_URL` into the system.

Desire 1: What is the canonical/pythonic/fast-ish-api way to test with a varied DB backend? Manual management of `$ENV` between testing and production? Config files - but how then do you switch the file it will read?

Desire 2: To configure my fastapi application at least somewhat at runtime so that my tests can control the backend, for example by parameterising fixtures for different DB engines. I have looked at an OO-way of instantiating the FastAPI app, but I ran into the at-import-time problem further down and ran out of documentation fast in that direction.

In my Perl days, a BEGIN block would have worked around this but Python doesn't do that sort of thing. It is my hope that I am being blind to a solution and that you guys have a better approach than this. Can you help?

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

3.9.7

**Comment by STeveShary** [comment]
@nerdstrike I can totally understand where you are coming from.  When I first started using this, I also thought: "Wait, how would you test this?"  The interesting thing that that with Python, it's pretty simple two write some config switching code.  I'll let you know what we do at my company (We have quite a few fast-api services in prod).

The approach that I take is to have a service that will return you a configuration dict.  Pretty simple.  We have UNIT_TESTING (default), DEV, CI, TEST, STAGING, and PROD.   When the application starts up, we have our database service reach out to the config service (through DI) to get the URL.   The config service checks for an environment variable.  If none is set, we assume that we are in UNIT_TESTING and proceed to handle that scenario.

All of our launch scripts to start the app locally (usually in a docker-compose.yml) set the environment variable.  We deploy our app using K8s and so we just make sure that the correct environment variable in our deploymennt yamls are set so we point at the correct stuff.

Here is an example config function:
```
import os

from typing import Optional

CONFIGS = {
    'UNIT_TESTING': {
        'environment': 'unit testing'
    },
    'DEV': {
        'environment': 'dev'
    },
    'TEST': {
        'environment': 'test'
    },
    'STAGING': {
        'environment': 'staging'
    },
    'PROD': {
        'environment': 'prod'
    }
}


def get_config_value(key: str) -> Optional:
    """
    Function that will fetch a config value.  It will pull the value based on the 'APP_ENV' env variable
    that is set.  This should be set on the docker container when running.
    """
    env_config = CONFIGS[os.environ.get('APP_ENV', 'UNIT_TESTING')]
    return env_config[key]
```

**Comment by nerdstrike** [comment]
@STeveShary - indeed. If you operate in containers this sort of thing is suitable. The container always needs to know this by one method or another, e.g. via config file of old, or a mounted `.env` for `uvicorn`. I'd like my tests to "do the safe thing" if run on installation via pip as the production environment is quite anti-container. 🤔

**Comment by STeveShary** [comment]
@nerdstrike We have the same idea.  We have the setting that you _must_ specify an environment, otherwise the config assumes a unit testing environment.  That is why we have this line:
```python3
    env_config = CONFIGS[os.environ.get('APP_ENV', 'UNIT_TESTING')]
```

If no environment is provided, it will use `UNIT_TESTING`.

**Comment by grillazz** [comment]
@nerdstrike you can use make and docker compose i.e.
```make
.PHONY: test
test:	## Run project tests
	docker-compose -f docker-compose.yml -f docker-compose.test.yml  run --rm app pytest
```
and in docker-compose.test.yml you can do something like
```compose
version: '3'

services:
  app:
    environment:
      - SQL_DB=testdb

  db:
    environment:
      - POSTGRES_USER=${SQL_USER}
      - SQL_DB=testdb
```
my conftest.py bewlo:
```python
import pytest
from httpx import AsyncClient
from the_app.main import app
from the_app.models.base import Base
from the_app.database import engine


@pytest.fixture(
    params=[
        pytest.param(("asyncio", {"use_uvloop": True}), id="asyncio+uvloop"),
    ]
)
def anyio_backend(request):
    return request.param


async def start_db():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)
    # for AsyncEngine created in function scope, close and
    # clean-up pooled connections
    await engine.dispose()


@pytest.fixture
async def client() -> AsyncClient:
    async with AsyncClient(
        app=app,
        base_url="http://testserver/v1",
        headers={"Content-Type": "application/json"},
    ) as client:
        await start_db()
        yield client
        # for AsyncEngine created in function scope, close and
        # clean-up pooled connections
        await engine.dispose()
```
and database.py
```python
from typing import AsyncGenerator

from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

from the_app import config

global_settings = config.get_settings()
url = global_settings.asyncpg_url

engine = create_async_engine(
    url,
    future=True,
    echo=True,
)

# expire_on_commit=False will prevent attributes from being expired
# after commit.
async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)


# Dependency
async def get_db() -> AsyncGenerator:
    session = async_session()
    try:
        yield session
        await session.commit()
    except SQLAlchemyError as ex:
        await session.rollback()
        raise ex
    finally:
        await session.close()
```

**Comment by nerdstrike** [comment]
In the absence of a more elegant solution, I have settled for a "MODE" $var that the DB singleton consults when configuring. I'd prefer not to lean on container tech to make up for limitations brought about by import-time server configuration, but this approach can apply there too.
